File: Programmers/[87946] 피로도.java
package wk12;

class Solution {

    static int maxCount = -987654321;
    static int[][] dungeons;
    static int dungeonCount;
    static boolean[] isVisited;


    // 탐험한 던전 수, 현재 피로도
    static void dfs(int count, int current) {
        maxCount = Math.max(count, maxCount);
        if (current < 1) {
            return;
        }

        for (int i = 0; i < dungeonCount; i++) {
            if (!isVisited[i] && current >= dungeons[i][0]) {
                isVisited[i] = true;
                dfs(count + 1, current - dungeons[i][1]);
                isVisited[i] = false;
            }
        }
    }

    public int solution(int k, int[][] dungeons_) {
        dungeons = dungeons_;
        dungeonCount = dungeons_.length;
        isVisited = new boolean[dungeonCount];

        dfs(0, k);

        return maxCount;
    }
}

File: Programmers/[87946] 피로도.py
max_count = 0
is_visited = []

# 현재 피로도, 던전 정보, 돌은 던전 수
def dfs(current_cost, dungeons, count):
    global max_count
    max_count = max(count, max_count)

    for index in range(len(dungeons)):
        if not is_visited[index]:
            min_cost, cost = dungeons[index]

            if current_cost >= min_cost:
                is_visited[index] = True
                dfs(current_cost - cost, dungeons, count + 1)
                is_visited[index] = False


def solution(k, dungeons):
    global is_visited
    is_visited = [False for _ in range(len(dungeons))]

    dfs(k, dungeons, 0)

    return max_count
